lambda/tmc.ml

```diff
@@ -0,0 +1,854 @@
+        Constr.delay_impure ~block_id con @@ fun con ->
+        d.code ~tail ~dst ~delayed:(con :: delayed));
+      delayed_use_count = d.delayed_use_count;
+    }
+end
+
+(** The TMC transformation requires information flows in two opposite
+    directions: the information of which callsites can be rewritten in
+    destination-passing-style flows from the leaves of the code to the
+    root, and the information on whether we remain in tail-position
+    flows from the root to the leaves -- and also the knowledge of
+    which version of the function we currently want to generate, the
+    direct version or a destination-passing-style version.
+
+    To clarify this double flow of information, we split the TMC
+    transform in two steps:
+
+    1. A function [choice t] that takes a term and processes it from
+    leaves to root; it produces a "code choice", a piece of data of
+    type [lambda Choice.t], that contains information on how to transform the
+    input term [t] *parameterized* over the (still missing) contextual
+    information.
+
+    2. Code-production operators that have contextual information
+    to transform a "code choice" into the final code.
+
+    The code-production choices for a single term have type [lambda Choice.t];
+    using a parametrized type ['a Choice.t] is useful to represent
+    simultaneous choices over several subterms; for example
+    [(lambda * lambda) Choice.t] makes a choice for a pair of terms,
+    for example the [then] and [else] cases of a conditional. With
+    this parameter, ['a Choice.t] has an applicative structure, which
+    is useful to write the actual code transformation in the {!choice}
+    function.
+*)
+module Choice = struct
+  type 'a t = {
+    dps : 'a Dps.t;
+    direct : unit -> 'a;
+    has_tmc_calls : bool;
+    benefits_from_dps: bool;
+    explicit_tailcall_request: bool;
+  }
+  (**
+     An ['a Choice.t] represents code that may be written
+     in destination-passing style if its usage context allows it.
+     More precisely:
+
+     - If the surrounding context is already in destination-passing
+       style, it has a destination available, we should produce the
+       code in [dps] -- a function parametrized over the destination.
+
+     - If the surrounding context is in direct style (no destination
+       is available), we should produce the fallback code from
+       [direct].
+
+      (Note: [direct] is also a function (on [unit]) to ensure that any
+      effects performed during code production will only happen once we
+      do know that we want to produce the direct-style code.)
+
+     - [has_tmc_calls] is true when there are TMC opportunities
+       in the subterm -- if some calls are in tail-modulo-cons
+       position and are rewritten into tailcalls in the [dps] version.
+
+     - [benefits_from_dps] is true when the [dps] calls strictly more
+       TMC functions than the [direct] version. See the
+       {!choice_makeblock} case.
```

Comment:
We could be more precise, e.g. "In this case, when encountering a constructor, we want to create a destination and switch to a destination-passing style context even if no destination is currently available (see the {!choice_makeblock} case)."